src/front/spv/mod.rs

```diff
@@ -2098,66 +2486,137 @@ impl<I: Iterator<Item = u32>> Parser<I> {
-                        value: Some(value_lexp.handle),
-                    };
+                    let value_handle = get_expr_handle!(value_id, value_lexp);
+                    break Some(crate::Statement::Return {
+                        value: Some(value_handle),
+                    });
                 }
                 Op::Branch => {
                     inst.expect(2)?;
                     let target_id = self.next()?;
-                    break Terminator::Branch { target_id };
+
+                    if let Some(info) = block_ctx.mergers.get(&target_id) {
+                        block.extend(emitter.finish(expressions));
+                        block_ctx.blocks.insert(block_id, block);
+                        let body = &mut block_ctx.bodies[body_idx];
+                        body.push(Body::BlockId(block_id));
+
+                        match *info {
+                            MergeBlockInformation::LoopContinue => body.push(Body::Continue),
+                            MergeBlockInformation::LoopMerge
+                            | MergeBlockInformation::SwitchMerge => body.push(Body::Break),
+                            MergeBlockInformation::SelectionMerge => {}
+                        }
+
+                        return Ok(());
+                    }
+
+                    block_ctx.info.entry(target_id).or_insert(body_idx);
+
+                    break None;
                 }
                 Op::BranchConditional => {
+                    fn merger(body: &mut Vec<Body>, info: &MergeBlockInformation) {
+                        match *info {
+                            MergeBlockInformation::LoopContinue => body.push(Body::Continue),
+                            MergeBlockInformation::LoopMerge
+                            | MergeBlockInformation::SwitchMerge => body.push(Body::Break),
+                            MergeBlockInformation::SelectionMerge => {}
+                        }
+                    }
+
                     inst.expect_at_least(4)?;
 
-                    let condition_id = self.next()?;
-                    let condition = self.lookup_expression.lookup(condition_id)?.handle;
+                    let condition = {
+                        let condition_id = self.next()?;
+                        let lexp = self.lookup_expression.lookup(condition_id)?;
+                        get_expr_handle!(condition_id, lexp)
+                    };
 
                     let true_id = self.next()?;
                     let false_id = self.next()?;
 
-                    break Terminator::BranchConditional {
+                    let accept = block_ctx.bodies.len();
+                    let mut accept_block = Vec::new();
+
+                    if let Some(info) = block_ctx.mergers.get(&true_id) {
+                        merger(&mut accept_block, info)
+                    }
+
+                    block_ctx.bodies.push(accept_block);
+                    block_ctx.info.entry(true_id).or_insert(accept);
```

Comment:
I agree this should be moved to an `else` to reduce allocations and make the context dump cleaner, but `insert().unwrap()` doesn't do what you think, it will panic unless there was a previous entry.